Default cookies_serializer to :json

albertoalmagro opened the pull request:
### Summary

Rails' default `cookies_serializer` is set to `:json` since Rails 4.2 by configuration at `config/initializers/cookies_serializer.rb`. However if no value is provided, it defaults to `:marshal`. While this was OK 4 years ago when this was the default value until Rails 4.1, this should default to `:json` from now on as it is the true default value for Rails. If a change is needed, it is already posible to change the default value in the initializer.

### Additional information

I also plan to remove `config/initializers/cookies_serializer.rb` in further PRs, but for now I prefer to leave it so that users can easily find where to change `cookies_serializer`. In case you have a different opinion, please tell me if you prefer me to remove it already within this patch.

albertoalmagro committed:
Default cookies_serializer to :json

Rails' default `cookies_serializer` is set to `:json` since Rails 4.2 by
configuration at `config/initializers/cookies_serializer.rb`. However if
no value is provided, it defaults to `:marshal`.

Instead of always setting the cookies serializer with an initializer, this
should default to `:json` from now on. To opt-in and have `:json` as your default
cookies serializer set `action_dispatch.use_json_as_default_cookies_serializer = true`.

In Rails 6.1 the default cookies serializer will be set to `:json`.

albertoalmagro commented:
This PR is ready for review. Thanks in advance for your time reviewing it 😉

oniofchaos commented:
Looks fine to me. My original thought was it might need a deprecation notice, but if the default has already been JSON for a while then it seems fine. I would also lean towards one change at a time and remove the file entirely in another PR later.

albertoalmagro commented:
r? @rafaelfranca

albertoalmagro commented:
CR please 🙏

albertoalmagro commented:
CR please 🙏

albertoalmagro commented:
Could I get any feedback here before fixing conflicts in the specs? This PR has been here for 5 months without getting any feedback... 🙁

rafaelfranca commented:
Before changing the default can we change that value in the 6.0 defaults and deprecate when that config is nil? And only them, we can change the default in 6.1.

albertoalmagro commented:
I'm working on this new framework default value. I expect to have it ready tomorrow.

albertoalmagro commented:
@rafaelfranca I've added a new framework default for 6.0 as you requested. Could you please review again?

Thanks in advance for your time.

albertoalmagro commented:
@rafaelfranca @kaspth fixed conflicts, this patch is ready again.

kaspth commented:
Hey! Sorry for not getting to this. This is unfortunately one of the backwards compatibly tricky parts of the code, so I didn't want to merge it just before rc1. That's also generally why we've stayed about from reviewing this. There's probably some lessons embedded in the comments for https://github.com/rails/rails/pull/32937, if you want to take a look at that.

I don't understand the purpose of `default_cookies_serializer`. Now we have `cookies_serializer` and `default_cookies_serializer` which do the same thing. I find that it also lets too much credence to Marshal as a valid value.

How about `# Rails.application.config.cookies.serialize_with_json = true`? That shows our direction more clearly. It should have the same caveats as the other cookies configs.

I'll be out for RailsConf next week, so I probably won't check back until after. This is way overdue though and I appreciate you moving us forward! 😄

albertoalmagro commented:
Hi @kaspth! Test suite is back to green after last rebase. Thanks for your input! I like your suggestion because as you say, it shows our intention more clearly. However, I would like to keep pointing out that we are only changing the default serializer, so how about including `default` in that variable? Three options come to my mind, which one would you prefer?
- `Rails.application.config.cookies.serialize_with_json_by_default = true`
- `Rails.application.config.cookies.use_json_as_default_serializer = true`
- `Rails.application.config.cookies.serialize_by_default_with_json = true`

Hope you are enjoying RailsConf!

albertoalmagro commented:
Fixed conflicts in `CHANGELOG`.

@kaspth may I know your preference regarding the options I wrote in my last comment? Thanks!

albertoalmagro commented:
Ping.

albertoalmagro commented:
Any thoughts? @kaspth 

Thanks.

albertoalmagro commented:
@kaspth @rafaelfranca I have rebased and squashed all commits. I also took the decision to rename the config variable to `use_json_as_default_cookies_serializer` following the line suggested by @kaspth. Hope you like it.

Thanks for your review in advance.

albertoalmagro commented:
Rebased and fixed conflicts again... is there a chance this gets completely reviewed? This PR has been waiting almost for a year... 🙁  Thanks for your consideration.

albertoalmagro commented:
Changes done after Rails 6.0:
- [x] Created `railties/lib/rails/generators/rails/app/templates/config/initializers/new_framework_defaults_6_1.rb.tt` to store new configuration options.
- [x] Updated `CHANGELOG` to point to Rails 6.2 for the deprecation.

kaspth commented:
Hey hey, just wanted to briefly state why I bowed out here. I didn't feel that my last comment https://github.com/rails/rails/pull/33967#issuecomment-486813472 was fully addressed. I'm talking about the first paragraph with the call for trying to level up skills around cookies. Cookies are a notoriously tough area due to cross version serialization. When the progress and intrinsic push for this volunteered PR seemed to burn out and just became "review, plz" over and over, it felt like I may as well just write the whole thing myself and then the benefit of "fixing" this default just didn't seem that beneficial. I took it as a red flag that, despite my negligence and I am sorry about that, no further knowledge or improvement was being added based on self-discovery in the extra time. I expected more experience/self-drive from a volunteered PR in this tricky area but I wasn't clear and upfront about that, sorry about that. ❤️

albertoalmagro commented:
Hey @kaspth, thanks for commenting, the reality is that this wasn't the only PR I had waiting for review, so it reached a point in which I couldn't afford investing more time in PRs that didn't get any attention.

> I expected more experience/self-drive from a volunteered PR in this tricky area but I wasn't clear and upfront about that

If you can explain this point a little further it might be helpful for me in upcoming PRs. I believe I have always addressed all the feedback I received and in cases I didn't get any feedback I made my own decisions and then asked again for feedback to prove my them. If this was my project I would have the final word, but as it isn't the case and cookies as you well said are critical, feedback from the core team was always necessary.
